cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/BrokerMetricSample.java

```diff
@@ -19,23 +19,44 @@
  * The class hosting all the broker level metrics in {@link KafkaMetricDef}.
  */
 public class BrokerMetricSample extends MetricSample<String, BrokerEntity> {
-  private static final byte CURRENT_VERSION = 5;
+  static final byte MIN_SUPPORTED_VERSION = 4;
+  static final byte LATEST_SUPPORTED_VERSION = 5;
+  private final byte _deserializationVersion;
```

Comment:
Cruise Control always serializes the payload based on its latest known serialization format, and uses the deserialization version header to indicate the fields that have a meaningful value. 

To read a sample, Cruise Control should know if it is a version that its deserialzer supports or not. Here are some cases, where removing this version from the header would cause problems:

1. If the sample store topic keeps samples with `version-X`, and we run a Cruise Control instance, which can support the latest version of `version-(X-1)`, then if we do not know the deserialization version, then we will lose existing data. Note, critically, that losing existing data is different from backwards compatibility.
2. Today we retrieve the bytes to deserialize from a Kafka topic by consuming records from it. This design implicitly fits a single sample to a single record value. However, this is a pluggable component, which can use other means to store/retrieve data. Hence, the bytes to deserialize, which are transmitted from the network, might come with dummy data after the actual sample to be deserialized. Hence, without knowing the actual deserialization version, we are likely to read bad data and cannot distinguish it from corrupted data.

Reply:
Yes, I agree that `version` is a necessary field in byte stream. My original idea is to fill this field with `LATEST_SUPPORTED_VERSION` instead of  `_deserializationVersion`  since we always write in latest version format. 

But after rethinking about it, I think always writing in latest version format may not be the correct way. 
consider the scenario, we collect a `V4` sample from consuming a period of metric records, if we fill 0 to all `V5` fields and store this sample in `V5` version. later when we read sample store topic(like CC restarts) and get this message, we cannot tell these `v5` fields' 0 value is either because this field is missing or is the actual value(which happens to be 0)